Add `block.prevrandao` as alias for `block.difficulty` (#13759)
ethereum/solidity: Changelog.md

Compiler Features:
 * Commandline Interface: Return exit code ``2`` on uncaught exceptions.
 * Commandline Interface: Add `--no-cbor-metadata` that skips CBOR metadata from getting appended at the end of the bytecode.

>  * EVM: Basic support for the EVM version "Paris". The ``difficulty`` opcode was replaced by ``prevrandao`` and does not behave as before. It now returns a random number based on the randomness beacon provided by the beacon chain. For details, please see [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).

Reviewer: ```suggestion
 * EVM: Basic support for the EVM version "Paris". The ``difficulty`` opcode was replaced by ``prevrandao``. For details, please see [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
```
Should we have to explain EVM opcodes changes here? Asking because I am not aware of how we handle those kind of situations...

Author: Well, it actually changed the semantic of the opcode not only replace it. Both still available until `0.9`, but currently they work as `prevrandao`.

Maybe change it to something like:
```markdown
* EVM: Basic support for the EVM version "Paris". The ``difficulty`` opcode now behaves as ``prevrandao``,
   i.e., it returns a random number based on a randomness beacon provided by the beacon chain.
   For details, please see [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
```
About the explanation, I can remove it, but other changelog entries had some explanation of the changes as well, so maybe it is fine ;)

Author: Or:
```markdown
* EVM: Basic support for the EVM version "Paris". Added ``prevrandao`` opcode, which returns a random number
   based on the randomness beacon provided by the beacon chain.
   The semantic of the ``difficulty`` opcode changed, and it now behaves as ``prevrandao``.
   For details, please see [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
```

Author: I changed to the last option. Let me know if it should be removed after all.

Reviewer: I think it's best to others more experienced (@ekpyron @cameel) to chime in. I was more curious about it, but it looks good.

Reviewer 2: I agree with @matheusaaguiar. Changelog is for the compiler so it's only relevant what compiler changes we introduced. We can still add a short note about the EIP to explain why, but this is not the main point of this changelog entry.

I'd do it like this:
```
 EVM: Basic support for the EVM version "Paris". Make ``prevrandao`` a valid opcode name that refers to the same opcode as the existing ``difficulty``. Emit the new name when code is targetting "Paris" and later EVM versions. The change is due to renaming introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
```

>  * EVM: Basic support for the EVM version "Paris". Make ``prevrandao`` a valid opcode name that refers to the same opcode as the existing ``difficulty``. Emit the new name when the code targets "Paris" and later EVM versions. The change is due to the renaming introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).

Reviewer 3: ```suggestion
 * EVM: Support for the EVM version "Paris". Introduce ``block.prevrandao`` in Solidity and ``prevrandao()`` in inline assembly, resp. as Yul builtin, to replace ``block.difficulty`` (still allowed with a warning when targetting Paris or later) and ``difficulty()`` (disallowed when targetting Paris or later). Use ``prevrandao`` as opcode name and in evmasm, when the code targets "Paris" and later EVM versions. The change is due to the renaming introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
```
We should probably update this to reflect the current state of the change. I actually think, contrary to earlier suggestions, that it may make sense to split this up into several points like:

- Support for the EVM version "Paris".
- Introduce ``block.prevrandao`` and ``prevrandao()`` in inline assembly for EVM versions >= paris.
- Deprecate ``block.difficulty`` and disallow ``difficulty()`` in inline assembly for EVM versions >= paris.

Or something along those lines.

Compiler Features:
 * Commandline Interface: Return exit code ``2`` on uncaught exceptions.
 * Commandline Interface: Add `--no-cbor-metadata` that skips CBOR metadata from getting appended at the end of the bytecode.
 * Support for the EVM version "Paris".
 * Introduce ``block.prevrandao`` in Solidity and ``prevrandao()`` in inline assembly for EVM versions >= paris.

>  * Deprecate ``block.difficulty`` and disallow ``difficulty()`` in inline assembly for EVM versions >= paris. The change is due to the renaming introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).

Reviewer 3: Might be good to keep a category like ``EVM`` (even though that's a bit weird), just for keeping these grouped - since the Changelog is sorted alphabetically (we should generally try to keep it sorted alphabetically and will at the latest sort it before the actual release).

Author: Added

>  * EVM: Support for the EVM version "Paris".
>  * EVM: Introduce ``block.prevrandao`` in Solidity and ``prevrandao()`` in inline assembly for EVM versions >= paris.
>  * EVM: Deprecate ``block.difficulty`` and disallow ``difficulty()`` in inline assembly for EVM versions >= paris. The change is due to the renaming introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).

Reviewer 3: Technically, that'd be (just resorted)

```suggestion
 * EVM: Deprecate ``block.difficulty`` and disallow ``difficulty()`` in inline assembly for EVM versions >= paris. The change is due to the renaming introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
 * EVM: Introduce ``block.prevrandao`` in Solidity and ``prevrandao()`` in inline assembly for EVM versions >= paris.
 * EVM: Support for the EVM version "Paris".
```

;-). Not there isn't other unsorted ones right above, but it's always better to avoid having to resort this before the release, since that's just annoying on ``git blame``s and such.

Author: hahaha I confess I intentionally left the `Support...` part before the others for a better context of the changes. But ok, I will respect the alphabetic order also for the text after the category. ;)

 * SMTChecker: Support Eldarica as a Horn solver for the CHC engine when using the CLI option ``--model-checker-solvers eld``. The binary `eld` must be available in the system.
 * SMTChecker: Make ``z3`` the default solver for the BMC and CHC engines instead of all solvers.
 * Parser: More detailed error messages about invalid version pragmas.

>  * Add support for EIP-4399 replacing ``difficulty`` by the ``prevrandao`` opcode which was introduced in EVM version "Paris".

Reviewer 4: I think this can be merged in the previous "Basic support" line.

 * SMTChecker: Make ``z3`` the default solver for the BMC and CHC engines instead of all solvers.
 * Parser: More detailed error messages about invalid version pragmas.

> 

Reviewer: There should be 2 lines between sections.
